src/expand/mod.rs

```diff
@@ -0,0 +1,622 @@
+        Regex::new(&re_str).map_err(Into::into)
+    }
+
+    /// Adds [`ParametersProvider`] for transforming into [`Regex`].
+    pub fn with_parameters<P: ParametersProvider<Spanned<'s>>>(
+        self,
+        parameters: P,
+    ) -> WithParameters<Self, P> {
+        WithParameters {
+            item: self,
+            parameters,
+        }
+    }
+}
+
+/// Possible errors while transforming `Input` representing
+/// [Cucumber Expression][0] into [`Regex`]
+///
+/// [0]: https://github.com/cucumber/cucumber-expressions#readme
+#[derive(Clone, Debug, Display, Error, From)]
+pub enum ExpansionError<Input>
+where
+    Input: fmt::Display,
+{
+    /// Parsing error.
+    #[display(fmt = "Parsing failed: {}", _0)]
+    Parsing(parse::Error<Input>),
+
+    /// Expansion error.
+    #[display(fmt = "Regex expansion failed: {}", _0)]
+    Expansion(UnknownParameterError<Input>),
+
+    /// Regex error.
+    #[display(fmt = "Regex failed: {}", _0)]
+    Regex(regex::Error),
+}
+
+/// [`Parameter`] not found.
+#[derive(Clone, Copy, Debug, Display, Error)]
+#[display(fmt = "Parameter '{}' not found.", not_found)]
+pub struct UnknownParameterError<Input>
+where
+    Input: fmt::Display,
+{
+    /// [`Parameter`] not found.
+    pub not_found: Input,
+}
+
+/// Trait for converting [Cucumber Expressions][0] [AST] elements into
+/// [`Iterator`]`<`[`Item`]` = `[`char`]`>` for transforming into [`Regex`].
+///
+/// [0]: https://github.com/cucumber/cucumber-expressions#readme
+/// [AST]: https://en.wikipedia.org/wiki/Abstract_syntax_tree
+/// [`Item`]: Iterator::Item
+pub trait IntoRegexCharIter<Input: fmt::Display> {
+    /// [`Iterator`] for transforming into [`Regex`].
+    type Iter: Iterator<Item = Result<char, UnknownParameterError<Input>>>;
+
+    /// Returns [`Iterator`] of [`char`]s for transforming into [`Regex`].
+    fn into_regex_char_iter(self) -> Self::Iter;
+}
+
+impl<Input> IntoRegexCharIter<Input> for Expression<Input>
+where
+    Input: Clone + fmt::Display + InputIter,
+    <Input as InputIter>::Item: AsChar,
+{
+    type Iter = ExpressionIter<Input>;
+
+    fn into_regex_char_iter(self) -> Self::Iter {
+        let into_regex_char_iter: fn(_) -> _ =
+            IntoRegexCharIter::into_regex_char_iter;
+        iter::once(Ok('^'))
+            .chain(self.0.into_iter().flat_map(into_regex_char_iter))
+            .chain(iter::once(Ok('$')))
+    }
+}
+
+/// [`IntoRegexCharIter::Iter`] for [`Expression`].
+type ExpressionIter<Input> = iter::Chain<
+    iter::Chain<
+        iter::Once<Result<char, UnknownParameterError<Input>>>,
+        iter::FlatMap<
+            vec::IntoIter<SingleExpression<Input>>,
+            <SingleExpression<Input> as IntoRegexCharIter<Input>>::Iter,
+            fn(
+                SingleExpression<Input>,
+            )
+                -> <SingleExpression<Input> as IntoRegexCharIter<Input>>::Iter,
+        >,
+    >,
+    iter::Once<Result<char, UnknownParameterError<Input>>>,
```

Comment:
Once we have TAIT we won't need this aliases anymore. It's worth mentioning with TODO comment for future.